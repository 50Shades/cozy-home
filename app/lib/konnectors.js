/* konnector lib ready to be added to cozy-client-js */
const KONNECTORS_DOCTYPE = 'io.cozy.konnectors'

const STATE_READY = 'ready'

export function addAccount (cozy, konnector, account) {
  if (!konnector.accounts) konnector.accounts = []
  konnector.accounts.push(account)
  return Promise.resolve(konnector)
}

export function fetchManifest (cozy, source) {
  return cozy.fetchJSON('GET', `/konnectors/manifests?Source=${encodeURIComponent(source)}`)
}

let cachedSlugIndex
function getSlugIndex (cozy) {
  return cachedSlugIndex
    ? Promise.resolve(cachedSlugIndex)
      : cozy.data.defineIndex(KONNECTORS_DOCTYPE, ['slug'])
          .then(index => {
            cachedSlugIndex = index
            return index
          })
}

export function findBySlug (cozy, slug) {
  if (!slug) throw new Error('Missing `slug` parameter')

  return getSlugIndex(cozy)
    .then(index => cozy.data.query(index, {selector: {slug: slug}}))
    .then(list => list.length ? list[0] : null)
}

export function install (cozy, slug, source, timeout = 120000) {
  if (!slug) throw new Error('Missing `slug` paramater for konnector')
  if (!source) throw new Error('Missing `source` parameter for konnector')

  return cozy.fetchJSON('POST', `/konnectors/${slug}?Source=${encodeURIComponent(source)}`)
      .then(konnector => new Promise((resolve, reject) => {
        const idTimeout = setTimeout(() => {
          reject(new Error('Konnector installation timed out'))
        }, timeout)

        // monitor the status of the connector
        // TODO: replace by a polling abstraction utility.
        const idInterval = setInterval(() => {
          cozy.data.find(KONNECTORS_DOCTYPE, konnector._id)
            .then(konnector => {
              if (konnector.state === STATE_READY) {
                clearTimeout(idTimeout)
                clearInterval(idInterval)
                resolve(konnector)
              }
            })
            .catch(error => {
              clearTimeout(idTimeout)
              clearInterval(idInterval)
              reject(error)
            })
        }, 1000)
      }))
}
