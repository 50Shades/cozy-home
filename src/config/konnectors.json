[
  {
    "slug": "ameli",
    "name": "Ameli",
    "vendorLink": "https://www.ameli.fr/",
    "category": "insurance",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/briced/cozy-konnector-v3-ameli.git#build"
  },
  {
    "slug": "axabanque102",
    "name": "Axa Banque",
    "vendorLink": "https://www.axabanque.fr/",
    "category": "banking",
    "icon": "axa.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "102"
    }
  },
  {
    "slug": "bforbank97",
    "name": "BforBank",
    "vendorLink": "https://www.bforbank.com/",
    "category": "banking",
    "icon": "bforbank.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "dob": {
        "type": "date"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "97"
    }
  },
  {
    "slug": "bnpparibas82",
    "name": "BNP Paribas",
    "vendorLink": "https://mabanque.bnpparibas",
    "category": "banking",
    "icon": "bnpparibas.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "82"
    }
  },
  {
    "slug": "bouyguesbox",
    "name": "Bouygues Box",
    "vendorLink": "https://www.bouyguestelecom.fr/",
    "category": "isp",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-bouyguesbox.git#build"
  },
  {
    "slug": "bouyguestelecom",
    "name": "Bouygues Telecom",
    "vendorLink": "https://www.bouyguestelecom.fr/",
    "category": "telecom",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-bouyguestelecom.git#build"
  },
  {
    "slug": "caatlantica3",
    "name": "Crédit Agricole",
    "icon": "ca.png",
    "category": "banking",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      },
      "bankId": {
        "type": "dropdown",
        "label": "branchName",
        "options": [
          {
            "name": "Alsace-Vosges",
            "value": "3"
          },
          {
            "name": "Brie-Picardie",
            "value": "5"
          },
          {
            "name": "Franche Comté",
            "value": "18"
          },
          {
            "name": "Ile de France",
            "value": "20"
          },
          {
            "name": "Lorraine",
            "value": "24"
          },
          {
            "name": "Nord-Est",
            "value": "28"
          },
          {
            "name": "Pyrénées Gascogne",
            "value": "33"
          },
          {
            "name": "Val de France",
            "value": "40"
          },
          {
            "name": "Alpes Provence",
            "value": "2"
          },
          {
            "name": "Aquitaine",
            "value": "4"
          },
          {
            "name": "Atlantique-Vendée",
            "value": "6"
          },
          {
            "name": "Anjou-Maine",
            "value": "7"
          },
          {
            "name": "Centre Est",
            "value": "8"
          },
          {
            "name": "Centre France",
            "value": "9"
          },
          {
            "name": "Centre Loire",
            "value": "10"
          },
          {
            "name": "Centre Ouest",
            "value": "11"
          },
          {
            "name": "Champagne Bourgogne",
            "value": "12"
          },
          {
            "name": "Charente-Maritime Deux-Sèvres",
            "value": "13"
          },
          {
            "name": "Charente Périgord",
            "value": "14"
          },
          {
            "name": "Corse",
            "value": "15"
          },
          {
            "name": "Côtes d'Armor",
            "value": "16"
          },
          {
            "name": "Finistère",
            "value": "17"
          },
          {
            "name": "Guadeloupe",
            "value": "19"
          },
          {
            "name": "Ille-et-Vilaine",
            "value": "21"
          },
          {
            "name": "Languedoc",
            "value": "22"
          },
          {
            "name": "Loire Haute-Loire",
            "value": "23"
          },
          {
            "name": "Martinique",
            "value": "25"
          },
          {
            "name": "Morbihan",
            "value": "26"
          },
          {
            "name": "Nord de France",
            "value": "27"
          },
          {
            "name": "Nord Midi Pyrénées",
            "value": "29"
          },
          {
            "name": "Normandie",
            "value": "30"
          },
          {
            "name": "Normandie-Seine",
            "value": "31"
          },
          {
            "name": "Provence Côte d'Azur",
            "value": "32"
          },
          {
            "name": "Réunion",
            "value": "34"
          },
          {
            "name": "Des Savoie",
            "value": "35"
          },
          {
            "name": "Sud Méditerranée",
            "value": "36"
          },
          {
            "name": "Sud Rhône Alpes",
            "value": "37"
          },
          {
            "name": "Toulouse 31",
            "value": "38"
          },
          {
            "name": "Touraine-Poitou",
            "value": "39"
          }
        ]
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build"
  },
  {
    "slug": "caissedepargne1",
    "name": "Caisse d'épargne (Particuliers)",
    "vendorLink": "https://www.caisse-epargne.fr",
    "category": "banking",
    "icon": "caissedepargne.svg",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "1"
    }
  },
  {
    "slug": "cdngroup109",
    "name": "Société Marseillaise de Crédit (Particuliers)",
    "vendorLink": "https://www.smc.fr/",
    "icon": "smc.png",
    "category": "banking",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "109"
    }
  },
  {
    "slug": "cdngroup88",
    "name": "Crédit du Nord (Particuliers)",
    "vendorLink": "https://www.credit-du-nord.fr/",
    "icon": "creditdunord.png",
    "category": "banking",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "88"
    }
  },
  {
    "slug": "comptenickel168",
    "name": "Compte Nickel",
    "vendorLink": "https://compte-nickel.fr",
    "category": "banking",
    "icon": "comptenickel.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "168"
    }
  },
  {
    "name": "Debug",
    "slug": "debug",
    "vendorLink": "http://cozy.io",
    "category": "other",
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-debug.git#build"
  },
  {
    "slug": "digiposte",
    "name": "Digiposte",
    "vendorLink": "https://secure.digiposte.fr",
    "category": "other",
    "dataType": [
      "document"
    ],
    "fields": {
      "email": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-digiposte.git#build"
  },
  {
    "slug": "edf",
    "name": "EDF",
    "vendorLink": "https://www.edf.fr/",
    "category": "energy",
    "dataType": [
      "bill",
      "contract",
      "consumption",
      "profile",
      "home"
    ],
    "fields": {
      "email": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-edf.git#build"
  },
  {
    "slug": "free",
    "name": "Free",
    "vendorLink": "https://www.free.fr/",
    "category": "isp",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-free.git#build"
  },
  {
    "slug": "freemobile",
    "name": "Free Mobile",
    "vendorLink": "https://mobile.free.fr/",
    "category": "telecom",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-free-mobile.git#build"
  },
  {
    "slug": "harmonie",
    "name": "Harmonie Mutuelle",
    "vendorLink": "https://www.harmonie-mutuelle.fr/",
    "category": "insurance",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-harmonie.git#build"
  },
  {
    "slug": "hellobank145",
    "name": "Hello bank!",
    "vendorLink": "https://www.hellobank.fr",
    "category": "banking",
    "icon": "hellobank.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "145"
    }
  },
  {
    "slug": "ingdirect95",
    "name": "ING Direct",
    "vendorLink": "https://secure.ingdirect.fr/public/displayLogin.jsf",
    "category": "banking",
    "icon": "ingdirect.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "dob": {
        "type": "date"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "95"
    }
  },
  {
    "slug": "labanquepostale175",
    "name": "Banque Postale (Particuliers)",
    "vendorLink": "https://www.labanquepostale.fr",
    "category": "banking",
    "icon": "labanquepostale.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "175"
    }
  },
  {
    "slug": "labanquepostale44",
    "name": "La Banque Postale (Particuliers)",
    "vendorLink": "https://www.labanquepostale.fr",
    "category": "banking",
    "icon": "labanquepostale.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "44"
    }
  },
  {
    "slug": "lcl143",
    "name": "LCL (Particuliers)",
    "vendorLink": "https://particuliers.secure.lcl.fr",
    "category": "banking",
    "icon": "lcl.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "new_identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "143"
    }
  },
  {
    "slug": "lcl144",
    "name": "e.LCL",
    "vendorLink": "https://e.lcl.fr/",
    "category": "banking",
    "icon": "lcl.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "new_identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "144"
    }
  },
  {
    "slug": "lcl146",
    "name": "LCL (Professionnels)",
    "vendorLink": "https://professionnels.secure.lcl.fr",
    "category": "banking",
    "icon": "lcl.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "new_identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "146"
    }
  },
  {
    "oauth": true,
    "slug": "maif",
    "name": "MAIF",
    "vendorLink": "maif.fr",
    "category": "insurance",
    "fields": {
      "access_token": {
        "type": "hidden"
      }
    },
    "dataType": [
      "profile",
      "contract",
      "home",
      "family",
      "sinister"
    ],
    "source": "git://github.com/cozy/cozy-konnector-maif.git#build",
    "oauth_scope": "openid+profile+offline_access"
  },
  {
    "slug": "malakoffmederic",
    "name": "Malakoff Mederic",
    "vendorLink": "http://www.malakoffmederic.com/",
    "category": "insurance",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-malakoffmederic.git#build"
  },
  {
    "slug": "materielnet",
    "name": "Materiel.net",
    "vendorLink": "https://www.materiel.net/",
    "category": "shopping",
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "dataType": [
      "bill"
    ],
    "source": "git://github.com/nicofrand/cozy-konnector-materielnet.git#build"
  },
  {
    "slug": "monabanq96",
    "name": "Monabanq",
    "vendorLink": "https://www.monabanq.com/client/",
    "category": "banking",
    "icon": "monabanq.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "96"
    }
  },
  {
    "slug": "numericable",
    "name": "Numéricable",
    "vendorLink": "https://www.numericable.fr/",
    "category": "isp",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-numericable.git#build"
  },
  {
    "slug": "orange",
    "name": "Orange Factures",
    "vendorLink": "https://www.orange.fr/",
    "category": "telecom",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-orange.git#build"
  },
  {
    "oauth": true,
    "slug": "orangevideos",
    "name": "Orange Vidéos",
    "vendorLink": "https://www.orange.fr/",
    "category": "isp",
    "dataType": [
      "videostream"
    ],
    "fields": {
      "access_token": {
        "type": "hidden"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "additionnalSuccessMessage": {
      "message": "connector.orange.message.delay"
    },
    "hasDescriptions": {
      "service": true
    },
    "source": "git://github.com/cozy/cozy-konnector-orangevod.git#build",
    "oauth_scope": "I"
  },
  {
    "oauth": true,
    "slug": "orangemobile",
    "name": "Orange Mobile – Appels & Géoloc",
    "vendorLink": "https://www.orange.fr/",
    "category": "telecom",
    "dataType": [
      "geopoint",
      "phonecommunicationlog"
    ],
    "fields": {
      "access_token": {
        "type": "hidden"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      },
      "agreement": {
        "type": "checkbox",
        "hasDescription": true
      }
    },
    "additionnalSuccessMessage": {
      "message": "connector.orange.message.delay"
    },
    "hasDescriptions": {
      "service": true,
      "connector": true
    },
    "source": "git://gitlab.cozycloud.cc/gjacquart/cozy-konnector-orangemobile.git#build",
    "oauth_scope": "M"
  },
  {
    "slug": "redmobile",
    "name": "SFR RED mobile",
    "vendorLink": "https://www.red-by-sfr.fr/",
    "category": "telecom",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-redmobile.git#build"
  },
  {
    "slug": "sfrbox",
    "name": "SFR Box",
    "vendorLink": "espace-client.sfr.fr/facture-fixe/consultation",
    "category": "isp",
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "dataType": [
      "bill"
    ],
    "source": "git://github.com/cozy/cozy-konnector-sfrbox.git#build"
  },
  {
    "slug": "sfrmobile",
    "name": "SFR Mobile",
    "vendorLink": "espace-client.sfr.fr/facture-mobile/consultation",
    "category": "telecom",
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "dataType": [
      "bill"
    ],
    "source": "git://github.com/cozy/cozy-konnector-sfrmobile.git#build"
  },
  {
    "slug": "sncf",
    "name": "Voyage-Sncf",
    "vendorLink": "https://www.voyages-sncf.com/",
    "category": "transport",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-sncf.git#build"
  },
  {
    "slug": "societegenerale155",
    "name": "Société Générale (Professionnels)",
    "vendorLink": "https://professionnels.societegenerale.fr",
    "category": "banking",
    "icon": "societegenerale.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "155"
    }
  },
  {
    "slug": "societegenerale156",
    "name": "Société Générale (Entreprises)",
    "vendorLink": "https://entreprises.societegenerale.fr",
    "category": "banking",
    "icon": "societegenerale.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "156"
    }
  },
  {
    "slug": "societegenerale157",
    "name": "Société Générale (Associations)- Progéliance Net",
    "vendorLink": "https://professionnels.societegenerale.fr/association_connexion.html",
    "category": "banking",
    "icon": "societegenerale.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "157"
    }
  },
  {
    "slug": "societegenerale158",
    "name": "Société Générale (Associations) - Sogecash Net",
    "vendorLink": "https://entreprises.societegenerale.fr/associations-connexion.html",
    "category": "banking",
    "icon": "societegenerale.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "158"
    }
  },
  {
    "slug": "societegenerale42",
    "name": "Sociéte Générale (Particuliers)",
    "vendorLink": "https://particuliers.societegenerale.fr",
    "category": "banking",
    "icon": "societegenerale.png",
    "dataType": [
      "bankAccounts",
      "bankTransactions"
    ],
    "fields": {
      "identifier": {
        "type": "text"
      },
      "secret": {
        "type": "password"
      }
    },
    "source": "git+ssh://git@gitlab.cozycloud.cc/labs/cozy-konnector-linxo.git#build",
    "parameters": {
      "bankId": "42"
    }
  },
  {
    "slug": "sosh",
    "name": "Sosh",
    "vendorLink": "https://www.sosh.fr/",
    "category": "telecom",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-sosh.git#build"
  },
  {
    "slug": "trainline",
    "name": "Trainline",
    "vendorLink": "www.trainline.fr",
    "category": "transport",
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "dataType": [
      "bill"
    ],
    "source": "git://github.com/cozy/cozy-konnector-trainline.git#build"
  },
  {
    "slug": "virginmobile",
    "name": "Virgin Mobile",
    "vendorLink": "https://www.virginmobile.fr/",
    "category": "telecom",
    "dataType": [
      "bill"
    ],
    "fields": {
      "login": {
        "type": "text"
      },
      "password": {
        "type": "password"
      },
      "folderPath": {
        "type": "folder",
        "advanced": true
      }
    },
    "source": "git://github.com/cozy/cozy-konnector-virgin-mobile.git#build"
  }
]
