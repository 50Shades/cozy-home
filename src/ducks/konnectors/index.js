import { fetchCollection } from 'redux-cozy-client'

export const DOCTYPE = 'io.cozy.konnectors'

// CRUD

export const fetchKonnectors = () => fetchCollection('konnectors', DOCTYPE)

// Selectors

export const getKonnector = (state, slug) =>
  !!state.documents &&
  !!state.documents[DOCTYPE] &&
  state.documents[DOCTYPE][`${DOCTYPE}/${slug}`]

export const getKonnectorsByCategory = (state, category) =>
  !!state.documents &&
  !!state.documents[DOCTYPE] &&
  Object.keys(state.documents[DOCTYPE]).reduce((konnectors, slug) => {
    return state.documents[DOCTYPE][slug].category === category
      ? konnectors.concat([state.documents[DOCTYPE][slug]])
      : konnectors
  }, [])
