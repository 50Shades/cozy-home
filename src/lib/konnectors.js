/* konnector lib ready to be added to cozy-client-js */
const KONNECTORS_DOCTYPE = 'io.cozy.konnectors'

const STATE_READY = 'ready'

export function addAccount (cozy, konnector, account) {
  if (!konnector.accounts) konnector.accounts = []
  konnector.accounts.push(account)
  return Promise.resolve(konnector)
}

export function fetchManifest (cozy, source) {
  return cozy.fetchJSON('GET', `/konnectors/manifests?Source=${encodeURIComponent(source)}`)
}

let cachedSlugIndex
function getSlugIndex (cozy) {
  return cachedSlugIndex
    ? Promise.resolve(cachedSlugIndex)
      : cozy.data.defineIndex(KONNECTORS_DOCTYPE, ['slug'])
          .then(index => {
            cachedSlugIndex = index
            return index
          })
}

export function findBySlug (cozy, slug) {
  if (!slug) throw new Error('Missing `slug` parameter')

  return getSlugIndex(cozy)
    .then(index => cozy.data.query(index, {selector: {slug: slug}}))
    .then(list => list.length ? list[0] : null)
}

export function install (cozy, konnector, timeout = 120000) {
  ['slug', 'source'].forEach(property => {
    if (!konnector[property]) throw new Error(`Missing '${property}' property in konnector`)
  })

  const { _id, slug, source } = konnector

  return (_id
    ? cozy.data.find(KONNECTORS_DOCTYPE, _id)
      .catch(error => {
        console.debug('install error')
        if (error.status !== '404') throw error
        return null
      })
      : Promise.resolve(null))
    .then(konnector => konnector || cozy.fetchJSON('POST', `/konnectors/${slug}?Source=${encodeURIComponent(source)}`))
    .then(konnector => waitForKonnectorReady(cozy, konnector, timeout))
}

// monitor the status of the connector and resolve when the connector is ready
function waitForKonnectorReady (cozy, konnector, timeout) {
  return new Promise((resolve, reject) => {
    const idTimeout = setTimeout(() => {
      reject(new Error('Konnector installation timed out'))
    }, timeout)

    const idInterval = setInterval(() => {
      cozy.data.find(KONNECTORS_DOCTYPE, konnector._id)
        .then(konnectorResult => {
          if (konnectorResult.state === STATE_READY) {
            clearTimeout(idTimeout)
            clearInterval(idInterval)
            resolve(konnector)
          }
        })
        .catch(error => {
          clearTimeout(idTimeout)
          clearInterval(idInterval)
          reject(error)
        })
    }, 1000)
  })
}

export function run (cozy, konnector, account, timeout = 120 * 1000) {
  if (!konnector.attributes || !konnector.attributes.slug) {
    throw new Error('Missing `attributes.slug` parameter for konnector')
  }
  if (!account._id) throw new Error('Missing `_id` parameter for account')
  if (!account.folderId) throw new Error('Missing `folderId` parameter for account')

  return cozy.fetchJSON('POST', '/jobs/queue/konnector', {
    data: {
      attributes: {
        options: {
          priority: 10,
          timeout,
          max_exec_count: 1
        }
      },
      arguments: {
        konnector: konnector._id,
        account: account._id,
        folderToSave: account.folderId
      }
    }
  })
  .then(job => waitForJobFinished(cozy, job, timeout))
}

// monitor the status of the connector and resolve when the connector is ready
function waitForJobFinished (cozy, job, timeout) {
  return new Promise((resolve, reject) => {
    const idTimeout = setTimeout(() => {
      reject(new Error('Job timed out'))
    }, timeout)

    const idInterval = setInterval(() => {
      cozy.data.find(job._type, job._id)
        .then(job => {
          if (job.state === STATE_READY) {
            clearTimeout(idTimeout)
            clearInterval(idInterval)
            resolve(job)
          }
        })
        .catch(error => {
          clearTimeout(idTimeout)
          clearInterval(idInterval)
          reject(error)
        })
    }, 1000)
  })
}
