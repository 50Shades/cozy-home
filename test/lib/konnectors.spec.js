/* eslint-env jest */
import * as konnectors from '../../src/lib/konnectors'

// TODO: mutualize this code with jobs
// just to tests calling, results are tested in cozy-client-js
const cozyMock = {
  fetchJSON: jest.fn((method, endpoint, data) => Promise.resolve(data))
}

beforeEach(() => {
  // clear mocks calls
  cozyMock.fetchJSON.mockClear()
})

describe('konnectors lib', () => {
  describe('createTrigger', () => {
    const konnector = { slug: 'test' }
    const account = { _id: '963a51f6cdd34401b0904de32cc5578d' }
    const folder = { _id: 'daa147092e1c4a1da8c991cb2a194adc' }

    const options = {
      frequency: 'weekly',
      day: 1,
      hours: 14,
      minutes: 15
    }

    it('creates a trigger', () => {
      expect.assertions(1)
      return konnectors
        .createTrigger(cozyMock, konnector, account, folder, options)
        .then(data => expect(data).toMatchSnapshot())
    })

    it('creates a trigger without folder', () => {
      expect.assertions(1)
      return konnectors
        .createTrigger(cozyMock, konnector, account, null, options)
        .then(data => expect(data).toMatchSnapshot())
    })
  })
})
