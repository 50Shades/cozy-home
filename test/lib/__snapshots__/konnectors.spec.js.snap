exports[`konnectors lib createTrigger creates a trigger 1`] = `
Object {
  "data": Object {
    "attributes": Object {
      "arguments": "0 15 14 * * 1",
      "frequency": "weekly",
      "type": "@cron",
      "worker": "konnector",
      "worker_arguments": Object {
        "account": "963a51f6cdd34401b0904de32cc5578d",
        "folder_to_save": "daa147092e1c4a1da8c991cb2a194adc",
        "konnector": "test",
      },
    },
  },
}
`;

exports[`konnectors lib createTrigger creates a trigger without folder 1`] = `
Object {
  "data": Object {
    "attributes": Object {
      "arguments": "0 15 14 * * 1",
      "frequency": "weekly",
      "type": "@cron",
      "worker": "konnector",
      "worker_arguments": Object {
        "account": "963a51f6cdd34401b0904de32cc5578d",
        "konnector": "test",
      },
    },
  },
}
`;
